connect/src/routes/tokenBridge/automatic.ts

```diff
@@ -11,72 +12,104 @@ import { AutomaticRoute, ValidationResult } from "../route";
 export namespace AutomaticTokenBridgeRoute {
   export type Options = {
-    nativeGasDropoff?: bigint;
+    nativeGas?: bigint;
   };
 }
 
-export class AutomaticTokenBridgeRoute<N extends Network> extends AutomaticRoute<
-  N,
-  AutomaticTokenBridgeRoute.Options
-> {
+type Op = AutomaticTokenBridgeRoute.Options;
+export class AutomaticTokenBridgeRoute<N extends Network> extends AutomaticRoute<N, Op> {
   NATIVE_GAS_DROPOFF_SUPPORTED = true;
 
-  async isSupported(): Promise<boolean> {
+  static isSupported<N extends Network>(
+    fromChain: ChainContext<N>,
+    toChain: ChainContext<N>,
+  ): boolean {
     // No transfers to same chain
-    if (this.request.from.chain === this.request.to.chain) return false;
+    if (fromChain.chain === toChain.chain) return false;
 
     // No transfers to unsupported chains
-    const fromChain = this.wh.getChain(this.request.from.chain);
     if (!fromChain.supportsAutomaticTokenBridge()) return false;
-
-    const toChain = this.wh.getChain(this.request.to.chain);
     if (!toChain.supportsAutomaticTokenBridge()) return false;
 
-    const atb = await fromChain.getAutomaticTokenBridge();
-    if (isTokenId(this.request.source) && !atb.isRegisteredToken(this.request.source.address))
-      return false;
-
-    // If the destination token was set, and its different than what
-    // we'd get from a token bridge transfer, then this route is not supported
-    if (this.request.destination) {
-      const destToken = await TokenTransfer.lookupDestinationToken(
-        fromChain,
-        toChain,
-        this.request.source,
-      );
-      if (!isSameToken(destToken, this.request.destination)) return false;
-    }
-
     return true;
   }
+  static getDefaultOptions(): Op {
+    return { nativeGas: 0n };
+  }
 
   async isAvailable(): Promise<boolean> {
-    // TODO
+    const atb = await this.fromChain.getAutomaticTokenBridge();
+
+    if (isTokenId(this.request.source))
+      return await atb.isRegisteredToken(this.request.source.address);
+
     return true;
   }
 
-  getDefaultOptions(): AutomaticTokenBridgeRoute.Options {
-    return { nativeGasDropoff: 0n };
-  }
+  async validate(options?: Op): Promise<ValidationResult<Op>> {
+    options = options ?? AutomaticTokenBridgeRoute.getDefaultOptions();
 
-  async validate(options: AutomaticTokenBridgeRoute.Options): Promise<ValidationResult<Error>> {
-    const transfer = this.toTransferDetails(options);
     try {
-      await TokenTransfer.validateTransferDetails(this.wh, transfer);
-      return { valid: true };
+      const { amount, destination } = this.request;
+
+      const quote = await this.quote(options);
+      const { amount: fee } = quote.relayFee!;
+
+      if (amount < fee) throw new Error(`Amount must be greater than fee:  ${amount} > ${fee}`);
+
+      options.nativeGas = options.nativeGas ?? 0n;
+
+      if (destination) {
+        if (isTokenId(destination)) {
+          if (!isSameToken(quote.destinationToken.token, destination))
```

Comment:
I don't understand why you moved this out of `isSupported`. what is the purpose of `isSupported` now? and when is `validate` supposed to be called?

in Connect, `isSupported` and `isAvailable` are what is used to filter the routes so that the UI can show all of the available routes. a basic check like this is required to show whether TokenBridge is a supported route that the user can choose. pushing this into validate might make `isSupported` fast, but it also makes it pretty useless.

my conceptual understanding is this:

`isSupported`: can this route theoretically perform this transfer? this involves the selected chains, tokens, and amount.
`isAvailable`: check in with any online dependencies, like a relayer, to make sure they're repsonsive.
`validate`: once the user selects a supported & available route, they put in options. this checks if they are correct. the amount is not an option, it's a required input. same with the tokens.

Reply:
I'm thinking of `isSupported` as just a check to see if this route can even be created. If the protocols or chains passed are not supported, filter it out immediately.

I'm thinking of `isAvailable` as a filter on the input tokens. 

I'm thinking of `validate` as the last step before sending, modifying and passing back suggested options.


It could be that the methods aren't broken out perfectly but I think it does make sense to have at least 1 very fast check to filter out as many potential routes as possible before doing extra work like reaching out to ask the contract for details.

Reply:
yeah I'm open to any interface here, I literally copied this out of Connect. it's probably not ideal. all I know is, the UI has to not show routes that can't handle your inputs. in this case, you would show TokenBridge as a route before checking if the tokens are the same, because that check is inside `validate`. so that doesn't work.

> Maybe we do need something like isRouteSupported and isTokenPairSupported where the latter does the work of checking to see if its a valid path from src->dst tokens?

based on my read of Connect's code, we generally need to make sure of these things before showing a route as a choice in the UI:

- source and destination chains are supported
- source and destination tokens are supported

Reply:
>source and destination chains are supported

this can be `isSupported`

> source and destination tokens are supported

source token support can be `isAvailable` by checking acceptable input tokens

if the Destination token is set, some awkward logic associated:

If the user passes `native`, despite not having the ability to "swap" to anything, the AutoTB can do native gas dropoff. The `nativeGas` param in Options _must_ be set to one of `0n`/`undefined`/`amount - fee` to be sure no "extra" (non gas tokens) tokens are delivered. 

If the user passes anything else as a destination token, we can lookup the expected destination token and compare it. 

Both the fee and destination token are returned as part of fetching a quote.

> amount is not too small or too large

We can filter on truncated(amount) == 0 but we'll have to fetch a quote to determine if the fee is > amount. 

We may want to grab a quote early on, in `isAvailable` or some new async static initializer for example. 

The quote may be cache-able but should be considered invalid if the options are different than the last time we asked for a quote, in that case it needs to be re-fetched.

Reply:
I still don't understand the purpose of `isSupported` if all it does is check the two chains. you'll never just call that and be done; you always need to check more stuff before you can return a set of routes from the resolver.